### ISSUE TITLE
Appservice worker trips over itself during intermittent database problems.

### ISSUE THREAD

**Issue by turt2live** [matrix-org/synapse#2509]
*Note: all times are CEST, unless otherwise indicated*

**TLDR**: Appservice transactions can be sent out of order, reproduced by intermittent database connection problems. Reported because the message literally says "FIX ME".

A large spike in traffic caused synapse to exceed the maximum number of connections to postgres, yielding the following error in the appservice worker:
```
appservice - 2017-10-07 19:11:11,131 - synapse.appservice.scheduler - 157 - ERROR - - FATAL:  remaining connection slots are reserved for non-replication superuser connections
Traceback (most recent call last):
  File "/root/synapse/local/lib/python2.7/site-packages/synapse/appservice/scheduler.py", line 149, in send
    service_is_up = yield self._is_service_up(service)
  File "/root/synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/root/synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/root/synapse/local/lib/python2.7/site-packages/synapse/appservice/scheduler.py", line 193, in _is_service_up
    state = yield self.store.get_appservice_state(service)
  File "/root/synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/root/synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/root/synapse/local/lib/python2.7/site-packages/synapse/storage/appservice.py", line 218, in get_appservice_state
    desc="get_appservice_state",
  File "/root/synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/root/synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/root/synapse/local/lib/python2.7/site-packages/synapse/storage/_base.py", line 325, in runInteraction
    inner_func, *args, **kwargs
  File "/root/synapse/local/lib/python2.7/site-packages/twisted/python/threadpool.py", line 250, in inContext
    result = inContext.theWork()
  File "/root/synapse/local/lib/python2.7/site-packages/twisted/python/threadpool.py", line 266, in <lambda>
    inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
  File "/root/synapse/local/lib/python2.7/site-packages/twisted/python/context.py", line 122, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/root/synapse/local/lib/python2.7/site-packages/twisted/python/context.py", line 85, in callWithContext
    return func(*args,**kw)
  File "/root/synapse/local/lib/python2.7/site-packages/twisted/enterprise/adbapi.py", line 296, in _runWithConnection
    conn = self.connectionFactory(self)
  File "/root/synapse/local/lib/python2.7/site-packages/twisted/enterprise/adbapi.py", line 36, in __init__
    self.reconnect()
  File "/root/synapse/local/lib/python2.7/site-packages/twisted/enterprise/adbapi.py", line 76, in reconnect
    self._connection = self._pool.connect()
  File "/root/synapse/local/lib/python2.7/site-packages/twisted/enterprise/adbapi.py", line 431, in connect
    conn = self.dbapi.connect(*self.connargs, **self.connkw)
  File "/root/synapse/local/lib/python2.7/site-packages/psycopg2/__init__.py", line 130, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
OperationalError: FATAL:  remaining connection slots are reserved for non-replication superuser connections
```

As a result of that error, the worker started to go into recovery for all appservices (snippet of one below).
```
appservice - 2017-10-07 19:11:11,274 - synapse.appservice.scheduler - 185 - INFO - - Application service falling behind. Starting recoverer. AS ID e0a12be47d7cbf78a4abda31c35e769eed005738f30d3c6dfd23e9f4da796fe3
appservice - 2017-10-07 19:11:11,274 - synapse.appservice.scheduler - 175 - INFO - - New active recoverers: 1
appservice - 2017-10-07 19:11:11,277 - synapse.appservice.scheduler - 185 - INFO - - Application service falling behind. Starting recoverer. AS ID telematrix
appservice - 2017-10-07 19:11:11,277 - synapse.appservice.scheduler - 175 - INFO - - New active recoverers: 2
```

Due to the intermittent nature of the DB connection problem, some transactions were still able to get through while some were stuck in the recovery queue: 

```
appservice - 2017-10-07 19:11:14,011 - synapse.storage.appservice - 308 - ERROR - - appservice: Completing a transaction which has an ID > 1 from the last ID sent to this AS. We've either dropped events or sent it to the AS out of order. FIX ME. last_txn=1156185 completing_txn=1077333 service_id=e0a12be47d7cbf78a4abda31c35e769eed005738f30d3c6dfd23e9f4da796fe3
```

From the logs for appservice `e0a12be47d7cbf78a4abda31c35e769eed005738f30d3c6dfd23e9f4da796fe3` (Discord) it is clear that the transactions were sent out of order (note: timezone is UTC for this log):

```
node-pre-gyp verb matrix-appservice-bridge ::ffff:127.0.0.1 - - [07/Oct/2017:17:11:10 +0000] "PUT /transactions/1156184?access_token=<REDACTED> HTTP/1.1" 200 2 "-" "Synapse/0.23.1"
node-pre-gyp verb matrix-appservice-bridge ::ffff:127.0.0.1 - - [07/Oct/2017:17:11:10 +0000] "PUT /transactions/1156185?access_token=<REDACTED> HTTP/1.1" 200 2 "-" "Synapse/0.23.1"
node-pre-gyp verb matrix-appservice-bridge ::ffff:127.0.0.1 - - [07/Oct/2017:17:11:13 +0000] "PUT /transactions/1077333?access_token=<REDACTED> HTTP/1.1" 200 2 "-" "Synapse/0.23.1"
node-pre-gyp verb matrix-appservice-bridge ::ffff:127.0.0.1 - - [07/Oct/2017:17:11:14 +0000] "PUT /transactions/1077333?access_token=<REDACTED> HTTP/1.1" 200 2 "-" "Synapse/0.23.1"
node-pre-gyp verb matrix-appservice-bridge ::ffff:127.0.0.1 - - [07/Oct/2017:17:11:14 +0000] "PUT /transactions/1096330?access_token=<REDACTED> HTTP/1.1" 200 2 "-" "Synapse/0.23.1"
node-pre-gyp verb matrix-appservice-bridge ::ffff:127.0.0.1 - - [07/Oct/2017:17:11:14 +0000] "PUT /transactions/1096330?access_token=<REDACTED> HTTP/1.1" 200 2 "-" "Synapse/0.23.1"
node-pre-gyp verb matrix-appservice-bridge ::ffff:127.0.0.1 - - [07/Oct/2017:17:11:15 +0000] "PUT /transactions/1100499?access_token=<REDACTED> HTTP/1.1" 200 2 "-" "Synapse/0.23.1"
node-pre-gyp verb matrix-appservice-bridge ::ffff:127.0.0.1 - - [07/Oct/2017:17:11:15 +0000] "PUT /transactions/1156186?access_token=<REDACTED> HTTP/1.1" 200 2 "-" "Synapse/0.23.1"
node-pre-gyp verb matrix-appservice-bridge ::ffff:127.0.0.1 - - [07/Oct/2017:17:11:15 +0000] "PUT /transactions/1156187?access_token=<REDACTED> HTTP/1.1" 200 2 "-" "Synapse/0.23.1"
```

### Version information

- **Homeserver**: t2bot.io
- **Version**:        0.23.1
- **Install method**: pip
- **Platform**:       Host OS: Ubuntu 16.04, Container: Ubuntu 16.04

**Comment by turt2live** [comment]
It's also very possible that this is related to the federation meltdown in the last week

**Comment by turt2live** [comment]
I've not seen this ever since - closing
